[flashinfer] [kernel] support for fp8 kv cache for trtllm prefill attention

mxz297 opened the pull request:
Summary:
trtllm prefill attention does not support bf16 Q + fp8 kv. To enable fp8 kv using trtllm, we can add a dequant kernel before pretill attention, which creates a mock kv cache and mock block table, which include only tokens needed by prefills.

Since for prefill, the involved tokens are small compared to what KV cache can hold overall, this dequant kernel brings relatively small overhead and allows us to enable fp8 kv.

Test Plan:
AIME

Low
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-low_temp1.0_20250903_113216', 'metric': 0.5}]

Medium:
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-medium_temp1.0_20250903_111346', 'metric': 0.7416666666666667}]

High:
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-high_temp1.0_20250903_131208', 'metric': 0.9}]

Rollback Plan:

Differential Revision: D81623711

houseroad reviewed (approved):
Looks good to me. Can we create an issue, to have the trtllm kernel support fp8 kv by default, no need to have the additional conversion from us. (for tracking purpose)

houseroad commented:
Btw, fix the pre-commit (linter) please

mxz297 commented:
@houseroad  seems like there is a draft PR https://github.com/vllm-project/vllm/pull/23647 that can potentially get rid of the conversion . However, even so, the conversion may stay because it represents different compute precision and may have accuracy implications.

houseroad commented:
Yeah, I think it makes sense to have this to land. Fix the linter please. :-)

mxz297 commented:
the failed test is test_cutlass_mla_decode.py, however, my change is not relevant to MLA though (it is paged attention only)

elvischenv commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -692,12 +790,14 @@ def forward(
 
         # Insert FP8 quant for query
-        if attn_metadata.q_data_type == FP8_DTYPE:
+        if attn_metadata.q_data_type == FP8_DTYPE \
+            and not self.use_bf16q_fp8kv_attn:
```
Please do that inside the init of FlashInferMetadataBuilder
https://github.com/vllm-project/vllm/blob/fb1a8f932a6dabf19f890dd905ebe217539fd147/vllm/v1/attention/backends/flashinfer.py#L197-L205

mxz297 replied:
revised accordingly

elvischenv commented on vllm/envs.py:
```diff
@@ -1140,6 +1141,10 @@ def get_vllm_port() -> Optional[int]:
 
+    # If set to 1, when we use fp8 kv, we do not quantize Q to fp8
+    "VLLM_USE_TRTLLM_BF16Q_FP8KV_ATTENTION":
+    lambda: bool(int(os.getenv("VLLM_USE_TRTLLM_BF16Q_FP8KV_ATTENTION", "1"))),
```
Just to confirm, will it be set to 0 by default?

mxz297 replied:
We can set this to 0 by default, when accuracy is good with QKV FP8 attn.

elvischenv replied:
I suggest changing this to something like `VLLM_FLASHINFER_DISABLE_Q_QUANTIZATION`

mxz297 replied:
Revised accordingly. And set VLLM_FLASHINFER_DISABLE_Q_QUANTIZATION default to be false

elvischenv commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -793,19 +893,40 @@ def forward(
+                            layer._v_scale,
+                        ))
+                else:
+                    mock_kv_cache = None
+                    mock_block_table = None
```
Just use 
```suggestion
                    mock_kv_cache = kv_cache_permute
                    mock_block_table = block_tables_prefill
```

mxz297 replied:
revised accordingly. Thanks!

mxz297 commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -48,8 +50,86 @@
+             dequantized_vals.to(tl.bfloat16))
+
+
+def trtllm_prefill_attn_kvfp8_dequant(
```
i put this triton kernel in this file simply because there are already triton kernels in this file:

https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/flashinfer.py#L1016

mxz297 commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -692,12 +790,13 @@ def forward(
 
         # Insert FP8 quant for query
-        if attn_metadata.q_data_type == FP8_DTYPE:
+        if attn_metadata.q_data_type == FP8_DTYPE and not self.use_bf16q_fp8kv_attn:
```
sorry... i am not sure what you mean here. Here I tried to avoid quantize Q to FP8 which is happening in forward pass. I am not sure how it can be done in init of the FlashInferMetadataBuilder...

mxz297 replied:
i see. Revised accordingly

mxz297 commented:
> > > This is expected since we need another Flashinfer version update for the TRTLLM FP8 kernel with sinks.
> > 
> > 
> > I tried with flashinfer v0.3.1.
> 
> Expected. We need another release for trtllm-gen before that: https://github.com/flashinfer-ai/flashinfer/blob/7c587eb16c9236b7797525411fe8aedea0c9ec05/flashinfer/artifacts.py#L111-L120

I see. Thanks! I will try again when the needed trtllm-gen release is out and switch VLLM_USE_TRTLLM_BF16Q_FP8KV_ATTENTION's default to be false when E2E accuracy is good.

mxz297 committed:
support for fp8 kv cache for trtllm prefill attention (#24197)

Summary:
trtllm prefill attention does not support bf16 Q + fp8 kv. To enable fp8 kv using trtllm, we can add a dequant kernel before pretill attention, which creates a mock kv cache and mock block table, which include only tokens needed by prefills.

Since for prefill, the involved tokens are small compared to what KV cache can hold overall, this dequant kernel brings relatively small overhead and allows us to enable fp8 kv.

Test Plan:
AIME

Low
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-low_temp1.0_20250903_113216', 'metric': 0.5}]

Medium:
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-medium_temp1.0_20250903_111346', 'metric': 0.7416666666666667}]

High:
[{'eval_name': 'aime25', 'model_name': 'gpt-oss-120b-high_temp1.0_20250903_131208', 'metric': 0.9}]

Rollback Plan:

Rollback Plan:

Reviewed By: houseroad

Differential Revision: D81623711

Pulled By: mxz297

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 committed:
Fix lint

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 commented:
Revised to use env VLLM_FLASHINFER_DISABLE_Q_QUANTIZATION and default to be false

elvischenv commented:
Overall looks good to me, thanks. cc @ProExpertProg @mgoin to check if you have any comments for the env var or something else.

The last thing is I don't think we need so much unrelated formatting. It should be good to pass the pre-commit with original state(like my previous PR #23647). Could you try reverting those and running pre-commit locally?

mxz297 committed:
Add "VLLM_FLASHINFER_DISABLE_Q_QUANTIZATION" to env hash

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 committed:
Revert unnecessary lint change

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 committed:
Merge branch 'upstream-main' into export-D81623711

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 commented:
Added the new env to env hash and reverted most unrelated lint changes. My local linter still decided to make some changes, but the churn seems to be much smaller now.

mxz297 commented:
> > Expected. We need another release for trtllm-gen before that: https://github.com/flashinfer-ai/flashinfer/blob/7c587eb16c9236b7797525411fe8aedea0c9ec05/flashinfer/artifacts.py#L111-L120

By the way, do you know when the trtllm-gen release will be available?

mxz297 committed:
Merge branch 'upstream-main' into export-D81623711

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mgoin commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -48,8 +49,86 @@
+        (batch_idx * block_table_stride + mock_block_table_idx + 1) *
+        KV_CACHE_STRIDE + K_CACHE_STRIDE + tl.arange(0, K_CACHE_STRIDE))
+    tl.store(mock_kv_cache_ptr + mock_cache_offset,
+             dequantized_vals.to(tl.bfloat16))
```
ditto

mgoin commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -48,8 +49,86 @@
+    new_s = (batch_size * num_of_page_per_token + 1, s[1], s[2], s[3], s[4])
+    # mock kv cache contains just the pages needed by this prefill
+    mock_kv_cache = torch.empty(new_s,
+                                dtype=torch.bfloat16,
```
ditto

mgoin commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -805,11 +884,28 @@ def forward(
+                            layer._v_scale,
+                        ))
+                else:
+                    mock_kv_cache = kv_cache_permute
+                    mock_block_table = block_tables_prefill
```
nit: I don't like using the term "mock" here. I realize we need to keep around `kv_cache_permute` for decode so we can't just reuse the name, but it is misleading to pass into the kernel

mxz297 commented on vllm/v1/attention/backends/flashinfer.py:
```diff
@@ -48,8 +49,86 @@
+    dequantized_vals = fp8_vals.to(tl.float32) * k_scale_val
+    mock_cache_offset = (batch_idx * block_table_stride + mock_block_table_idx
+                         + 1) * KV_CACHE_STRIDE + tl.arange(0, K_CACHE_STRIDE)
+    tl.store(mock_kv_cache_ptr + mock_cache_offset,
+             dequantized_vals.to(tl.bfloat16))
```
@mgoin triton does not support template dtype, so it is difficult to support generic types here. Do you have some specific types that you want to support beyond bfloat16?

mxz297 committed:
Generalize to support dequant to bf16 anf fp16

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 committed:
Directly get dequantize dtype within kernel

Signed-off-by: Xiaozhu <mxz297@gmail.com>

mxz297 committed:
Merge branch 'upstream-main' into export-D81623711

Signed-off-by: Xiaozhu <mxz297@gmail.com>
